torch/csrc/generic/Tensor.cpp

```diff
@@ -506,25 +506,83 @@ static bool THPTensor_(_indexOnce)(PyObject *index, int &indexed_dim,
 }
 
 #ifndef TH_REAL_IS_HALF
-static bool THPTensor_(_checkAdvancedIndexing)(THPTensor *indexed, PyObject *arg) {
-  // Currently we only support the integer-array indexing strategy for advanced
-  // indexing - where we have ndim sequence/LongTensor arguments
 
+static bool THPTensor_(_checkBasicIntegerArrayIndexing)(THPTensor *indexed, PyObject *arg) {
   long ndim = THTensor_(nDimension)(LIBRARY_STATE indexed->cdata);
 
-  // Verify that all of the inputs are either Tensors or Sequences.
   if (PySequence_Check(arg) && PySequence_Size(arg) == ndim) {
+    THPObjectPtr fast = THPObjectPtr(PySequence_Fast(arg, NULL));
     for (Py_ssize_t i = 0; i < ndim; ++i) {
-      PyObject *item = PySequence_GetItem(arg, i);
+      PyObject *item = PySequence_Fast_GET_ITEM(fast.get(), i);
       if (!THPLongTensor_Check(item) && !PySequence_Check(item)) {
-        Py_DECREF(item);
         return false;
       }
-      Py_DECREF(item);
     }
     return true;
   }
   return false;
+}
+
+static bool THPTensor_(_checkAdvancedIndexing)(THPTensor *indexed, PyObject *arg) {
+  // Currently we only support two forms of advanced indexing:
+  //
+  // 1. "Basic Integer Array Indexing" the integer-array indexing strategy
+  // where we have ndim sequence/LongTensor arguments
+  // 2. Combining Advanced Indexing with ":", with the limitation that
+  // the advanced indexing dimensions must be adjacent, i.e.:
+  //
+  // x[:, :, [1,2], [3,4], :] --> valid
+  // x[:, [1,2], :, [3,4], :] --> not valid
+
+  // Verification, Step #1 -- ndim sequencers
+  if (THPTensor_(_checkBasicIntegerArrayIndexing)(indexed, arg)) return true;
+
+  // Verification, Step #2 -- at least one sequencer, all the rest are
+  // ':', can be less than ndim indexers, all sequencers adjacent
+
+  long ndim = THTensor_(nDimension)(LIBRARY_STATE indexed->cdata);
+  // TODO: should this be == ndim? --> for now, yes, but to support
+  // other things, no
+  if (PySequence_Check(arg) && PySequence_Size(arg) == ndim) {
```

Comment:
if your advanced indexer is of 3d, you can do x[idx, :] where x is 4d. in that case, the PySequence size is 2 right? but it's still valud